ui/src/chat-with-kg/App.tsx

```diff
@@ -1,35 +1,100 @@
+import useWebSocket, { ReadyState } from "react-use-websocket";
+import {
+  ConversationState,
+  WebSocketRequest,
+  WebSocketResponse,
+} from "./types/websocketTypes";
 
 const SEND_REQUESTS = true;
 
-const chatMessageObjects: ChatMessageObject[] = SEND_REQUESTS ? [] : [
-  {
-    id: 0,
-    type: "input",
-    sender: "self",
-    message:
-      "This is the first message which has decently long text and would denote something typed by the user",
-  },
-  {
-    id: 1,
-    type: "text",
-    sender: "bot",
-    message:
-      "And here is another message which would denote a response from the server, which for now will only be text",
-  },
-];
+const chatMessageObjects: ChatMessageObject[] = SEND_REQUESTS
+  ? []
+  : [
+      {
+        id: 0,
+        type: "input",
+        sender: "self",
+        message:
+          "This is the first message which has decently long text and would denote something typed by the user",
+      },
+      {
+        id: 1,
+        type: "text",
+        sender: "bot",
+        message:
+          "And here is another message which would denote a response from the server, which for now will only be text",
+      },
+    ];
+
+const URL =
+  import.meta.env.VITE_KG_CHAT_BACKEND_ENDPOINT ??
+  "ws://localhost:7860/text2text";
 
 function App() {
-  const [chatMessages, setChatMesages] = useState(chatMessageObjects);
-  const [loading, setLoading] = useState(false);
+  const [chatMessages, setChatMessages] = useState(chatMessageObjects);
+  const [conversationState, setConversationState] =
+    useState<ConversationState>("ready");
+  const { sendJsonMessage, lastMessage, readyState } = useWebSocket(URL);
+
+  useEffect(() => {
+    if (!lastMessage) {
+      return;
+    }
+
+    const websocketResponse = JSON.parse(lastMessage.data) as WebSocketResponse;
+    console.log(websocketResponse);
+
+    if (websocketResponse.type === "debug") {
+      console.log(websocketResponse.detail);
+    } else if (websocketResponse.type === "error") {
+      setConversationState("error");
+      console.error(websocketResponse.detail);
+    } else if (websocketResponse.type === "start") {
+      setConversationState("streaming");
+
+      setChatMessages((chatMessages) => [
+        ...chatMessages,
+        {
+          id: chatMessages.length,
+          type: "text",
+          sender: "bot",
+          message: "",
+        },
+      ]);
+    } else if (websocketResponse.type === "stream") {
+      setChatMessages((chatMessages) => {
+        const lastChatMessage = chatMessages[chatMessages.length - 1];
+        const rest = chatMessages.slice(0, -1);
+
+        return [
+          ...rest,
+          {
+            ...lastChatMessage,
+            message: lastChatMessage.message + websocketResponse.output,
+          },
+        ];
+      });
+    } else if (websocketResponse.type === "end") {
+      setConversationState("ready");
+    }
+  }, [lastMessage, setConversationState, setChatMessages]);
```

Comment:
hmm, did you really mean to use the setters as the 2nd arguments to useEffect here? Maybe that works, but for readability I'd suggest/prefer just using the values instead of the setters... :-)

Reply:
I added them for competition's sake since the hook also depends on them. Although they should never actually change

Reply:
yes, but I was saying this should use `chatMessages` and `conversationState`, NOT the setters for those...

Reply:
Ah right. My bad good catch.